dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java

```diff
@@ -574,7 +578,10 @@ public Void visitCoherence(Relation co) {
                     BooleanFormula pairingCond = bmgr.and(execPair, sameAddress);
                     BooleanFormula coF = forwardPossible ? edge.encode(xz) : bmgr.makeFalse();
                     BooleanFormula coB = backwardPossible ? edge.encode(zx) : bmgr.makeFalse();
-                    enc.add(bmgr.equivalence(pairingCond, bmgr.or(coF, coB)));
+                    // Coherence is not total for weak stores in PTX
+                    if(!x.is(Tag.PTX.WEAK) || !z.is(Tag.PTX.WEAK)) {
+                        enc.add(bmgr.equivalence(pairingCond, bmgr.or(coF, coB)));
+                    }
```

Comment:
We also need to make sure that we have coherences between `morally-strong` stores.
Since `morally-strong` is a derived relation, we have to somehow formulate the coherence condition inside the cat file.
Maybe something like this works:
```
// Make sure that all morally strong same-loc writes are related by co in either direction
empty ([W];morally-strong;[W] & loc) \ (co | co^-1)
```

Reply:
The current implementation forces to have one of the co-orders if one store is weak, but the other is not. I don't think this is correct since they might still not be morally strong.

Using an && in line 582 and adding the axiom that Thomas proposes should solve the problem AFAIS.

Reply:
Is it possible that utilizing `&&` might still include morally-weak writes, like two `st.rlx.cta` from different ctas?

Reply:
Do you even need the check at all anymore? With the new Axiom in cat, you force a coherence between all morally strong writes. Axiom 1
```
(* Axiom 1 Coherence *)
let cause-write = ([W]; cause; [W]) & loc
empty (cause-write \ co) as axiom-Coherence
```
forces coherences between causally related writes as well. This should cover all cases, so that your encoding doesn't need to enforce any coherence at all, does it? I can imagine that transitivity is not yet given, so you might need to add `let co = co+` in your cat.

Reply:
We definitely need line 583 for the other models. The axiom you propose + axiom 1 guarantee that anything that is causal or morally strong should be part of `co`. But we still need to tell the solver that "if the are not morally strong and they are not causal, then you are free not to add any `co` edge". Or am I missing something?

Reply:
> We definitely need line 583 for the other models. 

Of course, we would skip encoding the equivalence only for PTX.

> The axiom you propose + axiom 1 guarantee that anything that is causal or morally strong should be part of `co`. But we still need to tell the solver that "if the are not morally strong and they are not causal, then you are free not to add any `co` edge". Or am I missing something?

The solver is free to not choose a coherence, because he is unconstrained (we do not encode totality anymore).
However, the solver could also choose some coherence, which is wrong according to the documentation: 
"Two overlapping writes are unrelated in coherence order if they are in a data-race".
We would further need to add something like:
```
let data-race = (M x (W \ IW) | (W \ IW) x M) \ morally-strong \ cause
empty data-race & co
```

More generally, the only properties of `co` that we can enforce in the anarchic semantics are that `co` relates only same-address writes, that it is transitive, and that it is acyclic. All other properties are related to derived relations and hence need to be defined in the cat. We could, in principle, formulate the former properties also in the .cat so that the anarchic encoding is basically empty:
```
acyclic co // acyclicity
empty co \ (WxW & loc) // co only relates same-address writes
let co = co+ // transitivity
```

Reply:
The cause-write and morally-strong-write seems cover all cases. But isn't the `co`  in dartagnan transitive by default? Or we need to add it because PTX is kept out from this constraints.

Reply:
I'm not sure if you get transitivity for free. Suppose you had only weak stores (a)-(c) related as follows `a -cause-> b -poloc-> c`, then causality gives you `a -co-> b` and morally strong gives you `b -co-> c`. The pair `(a, c)` is neither morally strong nor causally related, yet you should have `a -co-> c` by transitivity.

The encoding of `co` is only transitive in the non-PTX models because of totality. This is because if you have `a -co-> b` and `b -co-> c`, the solver must (by totality) choose a coherence between `a` and `c`, however `c -co-> a` would lead to cycle, so the solver has to choose `a -co-> c`. Without totality, the solver could simple choose neither of the two options!
You could do this in the .cat though:
```
empty ([W];(morally-strong | cause)[W] & loc)+ \ (co | co^-1);
```
I think this will make sure that `co` also relates stores over transitive chains over `cause` and `morally-strong` relations. Interestingly, it will not force `co` to be transitive, so it might be kinda wrong (it could still be non-transitive on racy stores).

Reply:
This won't cover the case I mentioned above where you get a transitive edge by the interaction between causally related stores and morally-strong stores.
The easiest way should be to just define `let co = co+` in the beginning of the cat and then just do the two axioms without transitive closures, i.e.
```
empty (([W]; cause; [W]) & loc) \ co as axiom-Coherence
empty (([W]; morally-strong; [W]) & loc)+ \ (co | co^-1) as axiom-Coherence2
```
I think `id` is not necessary in either case, because in the former you would have a store causally related to itself (which is surely not allowed) and in the latter I think `morally-strong` should already make sure to never relate an event to itself (at least the documentation suggests this).

Reply:
Yeah, I forgot to remove the transitive closure there. It was meant to be
```
empty (([W]; morally-strong; [W]) & loc) \ (co | co^-1) as axiom-Coherence2

```